src/Catalyst.Core.Modules.Dfs/RandomWalk.cs

```diff
@@ -0,0 +1,147 @@
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.Dfs.CoreApi;
+using Common.Logging;
+using Lib.P2P;
+using MultiFormats;
+
+namespace Catalyst.Core.Modules.Dfs
+{
+    /// <summary>
+    ///   Periodically queries the DHT to discover new peers.
+    /// </summary>
+    /// <remarks>
+    ///   A backgroud task is created to query the DHT.  It is designed
+    ///   to run often at startup and then less often at time increases.
+    /// </remarks>
+    public sealed class RandomWalk : IService
+    {
+        private static ILog log = LogManager.GetLogger(typeof(RandomWalk));
+        private CancellationTokenSource _cancel;
+
+        /// <summary>
+        ///   The Distributed Hash Table to query.
+        /// </summary>
+        internal IDhtApi Dht { get; set; }
+
+        /// <summary>
+        ///   The time to wait until running the query.
+        /// </summary>
+        public TimeSpan Delay = TimeSpan.FromSeconds(5);
+
+        /// <summary>
+        ///   The time to add to the <see cref="Delay"/>.
+        /// </summary>
+        public TimeSpan DelayIncrement = TimeSpan.FromSeconds(10);
+
+        /// <summary>
+        ///   The maximum <see cref="Delay"/>.
+        /// </summary>
+        public TimeSpan DelayMax = TimeSpan.FromMinutes(9);
+
+        /// <summary>
+        ///   Start a background process that will run a random
+        ///   walk every <see cref="Delay"/>.
+        /// </summary>
+        public Task StartAsync()
+        {
+            if (_cancel != null)
+            {
+                throw new Exception("Already started.");
+            }
+
+            _cancel = new CancellationTokenSource();
+            _ = RunnerAsync(_cancel.Token);
+
+            log.Debug("started");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   Stop the background process.
+        /// </summary>
+        public Task StopAsync()
+        {
+            _cancel?.Cancel();
+            _cancel?.Dispose();
+            _cancel = null;
+
+            log.Debug("stopped");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///   The background process.
+        /// </summary>
+        private async Task RunnerAsync(CancellationToken cancellation)
+        {
+            while (!cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(Delay, cancellation);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)